polygeist/lib/polygeist/Passes/ParallelLICM.cpp

```diff
@@ -362,36 +625,254 @@ void moveParallelLoopInvariantCode(AffineParallelOp loop) {
+                           SmallVectorImpl<Operation *> &opsToMove) {
+  // Do not use walk here, as we do not want to go into nested regions and
+  // hoist operations from there. These regions might have semantics unknown
+  // to this rewriting. If the nested regions are loops, they will have been
+  // processed.
+  SmallPtrSet<Operation *, 8> willBeMoved;
+  for (Block &block : loop.getLoopBody()) {
+    for (Operation &op : block.without_terminator()) {
+      if (!canBeHoisted(op, loop, willBeMoved, aliasAnalysis))
+        continue;
+      opsToMove.push_back(&op);
+      willBeMoved.insert(&op);
+    }
+  }
+}
+
+static size_t moveLoopInvariantCode(LoopLikeOpInterface loop,
+                                    const AliasAnalysis &aliasAnalysis) {
+  Operation *loopOp = loop;
+  if (!isa<scf::ParallelOp, AffineParallelOp, AffineForOp>(loopOp))
+    return 0;
+
+  SmallVector<Operation *, 8> opsToMove;
+  collectHoistableOperations(loop, aliasAnalysis, opsToMove);
+  if (opsToMove.empty())
+    return 0;
+
+  LoopGuardBuilder::create(loop)->guardLoop();
+
+  size_t numOpsHoisted = 0;
+  for (Operation *op : opsToMove) {
     loop.moveOutOfLoop(op);
+    ++numOpsHoisted;
+  }
 
-  LLVM_DEBUG(loop.print(llvm::dbgs() << "\n\nModified loop:\n"));
+  return numOpsHoisted;
 }
 
 void ParallelLICM::runOnOperation() {
+  AliasAnalysis &aliasAnalysis = getAnalysis<AliasAnalysis>();
+
+  [[maybe_unused]] auto getParentFunction = [](LoopLikeOpInterface loop) {
+    Operation *parentOp = loop;
+    do {
+      parentOp = parentOp->getParentOp();
+    } while (parentOp && !isa<func::FuncOp>(parentOp));
+    assert(parentOp && "Failed to find parent function");
+    return parentOp;
+  };
+
   getOperation()->walk([&](LoopLikeOpInterface loop) {
     LLVM_DEBUG({
       llvm::dbgs() << "----------------\n";
       loop.print(llvm::dbgs() << "Original loop:\n");
       llvm::dbgs() << "\n";
+      llvm::dbgs() << "in function " << *getParentFunction(loop) << "\n";
     });
 
-    moveLoopInvariantCode(loop);
-    TypeSwitch<Operation *>((Operation *)loop)
-        .Case<scf::ParallelOp>(
-            [&](auto loop) { moveParallelLoopInvariantCode(loop); })
-        .Case<AffineParallelOp>(
-            [&](auto loop) { moveParallelLoopInvariantCode(loop); });
+    // First use MLIR LICM to hoist simple operations.
+    {
+      size_t numOpHoisted = moveLoopInvariantCode(loop);
+
+      LLVM_DEBUG({
+        if (numOpHoisted)
+          loop.print(llvm::dbgs() << "Loop after MLIR LICM:\n");
+        llvm::dbgs() << "\nHoisted " << numOpHoisted << " operation(s)\n";
+        llvm::dbgs() << "----------------\n";
+        llvm::dbgs() << "in function " << *getParentFunction(loop) << "\n";
+        assert(mlir::verify(getParentFunction(loop)).succeeded());
+      });
+    }
+
+    // Now use this pass to hoist more complex operations.
+    {
+      numOpHoisted = moveLoopInvariantCode(loop, aliasAnalysis);
```

Comment:
is it possible that ParallelLICM is called more than once?
if so, we should use a local variable here, and add to the global variable.

Reply:
ok